### PR TITLE
Refactor RateMap to use numpy functions

### LINKED ISSUE #1172: Refactor RateMap to use numpy functions and add functionality
The current [intervals.py](https://github.com/tskit-dev/msprime/blob/master/msprime/intervals.py#L29) is a minimal refactoring required to add a more sensible interface and deprecate the old RecombinationMap.  We should at least refactor the current code to use numpy APIs rather than (e.g.) Python bisect. Fleshing out some additional simple functionality would be good too (but optional).

@grahamgower, you've been involved in the recombination map code before - this is a clean slate now, any thoughts?

### PR DESCRIPTION
**Pull request by grahamgower**
This pr, plus #1185, should close #1172.

### CODE REVIEW

File: msprime/intervals.py [hunk 1]
```diff
@@ -347,3 +310,65 @@ def get_length(self):
...
+        chr1	82571	        2.082414	0.080572
+        chr1	88169	        2.081358	0.092229
+        chr1	254996	        3.354927	0.439456
+        chr1	564598	        2.887498	1.478148
+        ...
+        chr1	182973428	2.512769	122.832331
+        chr1	183630013	0.000000	124.482178
+
+    :param str filename: The name of the file to be parsed. This may be
+        in plain text or gzipped plain text.
```
Comment (jeromekelleher) [inline msprime/intervals.py:345]: Need a return docstring here.

File: msprime/intervals.py [hunk 2]
```diff
@@ -347,3 +310,65 @@ def get_length(self):
...
+    """
+    # fmt: on
+    positions = []
```
Comment (jeromekelleher) [inline msprime/intervals.py:348]: I was hoping we could use numpy.loadtxt here, and make this a little less crappy.

File: msprime/intervals.py [hunk 3]
```diff
@@ -347,3 +310,65 @@ def get_length(self):
...
+            # Rate is expressed in centimorgans per megabase, which
+            # we convert to per-base rates
+            rates.append(rate * 1e-8)
+        if rate != 0:
+            raise ValueError(
+                "The last rate provided in the recombination map must be zero"
+            )
+    finally:
+        f.close()
+    return RecombinationMap(positions, rates, map_start=map_start)
```
Comment (jeromekelleher) [inline msprime/intervals.py:374]: We should return a RateMap here, rather than a RecombationMap (which is going for full deprecation)

File: msprime/intervals.py [hunk 4]
```diff
@@ -81,60 +80,66 @@ def slice(self, start=None, end=None, trim=False):  # noqa: A003
...
+                if rate[-1] == 0:
+                    position[-1] = self.position[-1]
                 else:
-                    new_positions.insert(0, 0)
-                    new_rates.insert(0, 0.0)
-            if new_positions[-1] != positions[-1]:
-                new_positions.append(positions[-1])
-                new_rates.append(0)
-        return self.__class__(new_positions, new_rates[:-1])
+                    position.append(self.position[-1])
```
Comment (jeromekelleher) [inline msprime/intervals.py:134]: Can use np.append here.

File: msprime/intervals.py [hunk 5]
```diff
@@ -81,60 +80,66 @@ def slice(self, start=None, end=None, trim=False):  # noqa: A003
...
+            # The simplest way to do this is to convert the position and
+            # rate arrays into regular lists, then prepend and/or append.
+            position = list(position)
+            rate = list(rate)
+            # Prepend or extend zero-rate region at start of map.
+            if position[0] != 0:
+                if rate[0] == 0:
+                    position[0] = 0
+                else:
+                    position.insert(0, 0)
```
Comment (jeromekelleher) [inline msprime/intervals.py:127]: Can use np.insert here.

File: msprime/intervals.py [hunk 6]
```diff
@@ -81,60 +80,66 @@ def slice(self, start=None, end=None, trim=False):  # noqa: A003
...
         if trim:
-            new_positions = [pos - start for pos in new_positions]
+            position -= start
         else:
-            if new_positions[0] != 0:
-                if new_rates[0] == 0:
-                    new_positions[0] = 0
+            # For the trim=False case we need to zero out all rates outside the
+            # interval [start, end], but retain the sequence_length.
+            # The simplest way to do this is to convert the position and
```
Comment (jeromekelleher) [inline msprime/intervals.py:118]: I find it hard to believe we can't easily replicate with numpy operations. Isn't it just a straight replace of insert/append for numpy equivalents?

File: msprime/intervals.py [hunk 7]
```diff
@@ -214,66 +219,24 @@ def uniform_map(cls, length, rate, num_loci=None):
...
-                )
-        finally:
-            f.close()
-        return cls(positions, rates, map_start=map_start)
+        warnings.warn(
+            "RecombinationMap.read_hapmap() is deprecated. "
+            "Use msprime.read_hapmap() instead.",
+            DeprecationWarning,
+        )
+        return read_hapmap(filename)
```
Comment (jeromekelleher) [inline msprime/intervals.py:239]: So, this would need to ``return RecombinationMap(rate_map.position, list(rate_map.rate) + [0])`` so we've got the right types and don't break old code.

**Review by jeromekelleher — COMMENTED** [review]
Looks good, thanks @grahamgower. Some comments.

File: msprime/intervals.py (hunk 2 above)
Comment (grahamgower) [reply msprime/intervals.py:348]: Ah, yes. Somehow I forgot about that.

File: msprime/intervals.py (hunk 3 above)
Comment (grahamgower) [reply msprime/intervals.py:374]: There's the small issue of the `map_start` parameter. How do you want to handle that? Add that to the rate map?

File: msprime/intervals.py (hunk 6 above)
Comment (grahamgower) [reply msprime/intervals.py:118]: Yes, that will work. But numpy's append and insert functions copy the arrays, and we're already doing so many copies around this code...

File: msprime/intervals.py (hunk 3 above)
Comment (jeromekelleher) [reply msprime/intervals.py:374]: Ah, yes. Yes, I guess we should add it to the RateMap all right.

File: msprime/intervals.py (hunk 6 above)
Comment (jeromekelleher) [reply msprime/intervals.py:118]: :shrug: it's a corner case and array copies are surprisingly fast.

**Comment by grahamgower** [comment]
Ok, this is ready for another look @jeromekelleher.

**Comment by grahamgower** [comment]
The test failure is a bit strange. Seems unrelated to the changes here though, as the failing test isn't using a recombination map.

**Review by jeromekelleher — APPROVED** [review]
Looks great, thanks @grahamgower!

**Comment by jeromekelleher** [comment]
The test failure is weird all right. Let's see if it shows up in other PRs.

**Comment by jeromekelleher** [comment]
Huh, weird. Travis issue didn't show up in #1210

**Comment by grahamgower** [comment]
I squashed/rebased, so we can see if it happens again.

**Comment by jeromekelleher** [comment]
Dunno what happened there. Probably some weird thing because we didn't specify a random seed somewhere. Not worth worrying about.
